src/qutip_qip/decompose/_utility.py

```diff
@@ -29,3 +30,180 @@ def check_gate(gate, num_qubits):
+        if output_form == "index_values":
+            output = gray_code_sequence_as_array_indices
+        else:
+            gray_code_as_binary = []
+            binary_sequence_list = _binary_sequence(num_qubits)
+            for i in gray_code_sequence_as_array_indices:
+                gray_code_as_binary.append(binary_sequence_list[i])
+            output = gray_code_as_binary
+    return(output)
+
+
+def _gray_code_steps(index_of_state_1, index_of_state_2, num_qubits):
+    """ Finds the sequence mapping from state 1 to state 2.
+
+    State 1 and 2 define the basis vectors of non-trivial values in a two-level
+    unitary matrix. Here, the inputs are their respective indices in a quantum
+    gate's array.
+
+    This function finds the number of steps between both states when only 1 bit
+    can be changed at each step.
+    """
+    # check array values are not the same
+    try:
+        assert (index_of_state_1 != index_of_state_2)
+    except AssertionError:
+        raise IndexError("Both indices need to be different.")
+
+    # array index for original binary sequence
+    indices_of_array = []
+    for i in range(2**num_qubits):
+        indices_of_array.append(i)
+
+    # Check both indices could be indices for num_qubits array
+    try:
+        assert all(x in indices_of_array for x in [
+                index_of_state_1, index_of_state_2])
+    except AssertionError:
+        raise IndexError(
+            "At least one of the input indices is invalid.")
+
+    # get the basis vector strings
+    binary_sequence_list = _binary_sequence(num_qubits)
+    state_1_binary = binary_sequence_list[index_of_state_1]
+    state_2_binary = binary_sequence_list[index_of_state_2]
+
+    # compare binary sequence positions to gray code sequence
+    # finds number of steps neeeded to go from one state to another
+    gray_code_sequence = _gray_code_sequence(num_qubits)
+    state_1_gray_code_index = gray_code_sequence.index(state_1_binary)
+    state_2_gray_code_index = gray_code_sequence.index(state_2_binary)
+    num_steps_gray_code = state_2_gray_code_index - state_1_gray_code_index
+
+    # create a smaller gray code sequence between the two states of interest
+    gray_code = _gray_code_sequence(num_qubits)[
+        state_1_gray_code_index:state_2_gray_code_index+1]
+    return(gray_code, num_steps_gray_code)
+
+
+def gray_code_plot(index_of_state_1, index_of_state_2, num_qubits):
+    """ Plots the difference between each step of a gray code sequence.
+
+    Parameters
+    -----------
+    index_of_state_i: int
+        The non-trivial indices in a two-level unitary matrix gate.
+
+    num_qubits:
+        Number of qubits being acted upon by the quantum gate.
+    """
+    # This function is still incomplete.
+
+
+def gray_code_gate_info(index_of_state_1, index_of_state_2, num_qubits):
+    """ Finds information about control and targets for CNOT gate in a gray
+    code sequence.
+    """
```

Comment:
Could you give the parameters or an example of a valid input? I tried a few but always get an error 
```At least one of the input indices is invalid.```

Reply:
Ah, you already give one above!

Reply:
I resolved a previous conversation but[ here's a linked description ](https://github.com/qutip/qutip-qip/pull/86#discussion_r677942416)(don't know if you saw it yet).

Reply:
Here's a 4 qubit example : 
![image](https://user-images.githubusercontent.com/66048318/127335404-c45d4f59-fc50-4c0e-a632-8b979df58bf5.png)